File: off_chain_data/application-java/app/src/main/java/parser/NamespaceReadWriteSet.java
/*
 * Copyright IBM Corp. All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package parser;

import com.google.protobuf.InvalidProtocolBufferException;
import org.hyperledger.fabric.protos.ledger.rwset.NsReadWriteSet;
import org.hyperledger.fabric.protos.ledger.rwset.kvrwset.KVRWSet;

public interface NamespaceReadWriteSet {
    String getNamespace();
    KVRWSet getReadWriteSet() throws InvalidProtocolBufferException;
    NsReadWriteSet toProto();
}

File: off_chain_data/application-go/parser/namespaceReadWriteSet.go
package parser

import (
	"sync"

	"github.com/hyperledger/fabric-protos-go-apiv2/ledger/rwset"
	"github.com/hyperledger/fabric-protos-go-apiv2/ledger/rwset/kvrwset"
	"google.golang.org/protobuf/proto"
)

type NamespaceReadWriteSet struct {
	nsReadWriteSet *rwset.NsReadWriteSet
	readWriteSet   func() (*kvrwset.KVRWSet, error)
}

func parseNamespaceReadWriteSet(nsRwSet *rwset.NsReadWriteSet) *NamespaceReadWriteSet {
	result := &NamespaceReadWriteSet{nsRwSet, nil}
	result.readWriteSet = sync.OnceValues(result.unmarshalReadWriteSet)
	return result
}

func (p *NamespaceReadWriteSet) Namespace() string {
	return p.nsReadWriteSet.GetNamespace()
}

func (p *NamespaceReadWriteSet) ReadWriteSet() (*kvrwset.KVRWSet, error) {
	return p.readWriteSet()
}

func (p *NamespaceReadWriteSet) unmarshalReadWriteSet() (*kvrwset.KVRWSet, error) {
	result := &kvrwset.KVRWSet{}
	if err := proto.Unmarshal(p.nsReadWriteSet.GetRwset(), result); err != nil {
		return nil, err
	}

	return result, nil
}
